ui/stories/components/table.stories.js

```diff
@@ -0,0 +1,483 @@
+        return this.get('controller.sortDescending') ? sorted.reverse() : sorted;
+      }),
+    },
+  };
+};
+
+export const MultiRow = () => {
+  return {
+    template: hbs`
+      <h5 class="title is-5">Multi-row Table</h5>
+      <ListTable @source={{sortedShortList}} @sortProperty={{controller.sortProperty}} @sortDescending={{controller.sortDescending}} @class="is-striped" as |t|>
+        <t.head>
+          <t.sort-by @prop="name">Name</t.sort-by>
+          <t.sort-by @prop="lang">Language</t.sort-by>
+        </t.head>
+        <t.body @key="model.name" as |row|>
+          <tr>
+            <td>{{row.model.name}}</td>
+            <td>{{row.model.lang}}</td>
+          </tr>
+          <tr>
+            <td colspan="2">{{row.model.desc}}</td>
+          </tr>
+        </t.body>
+      </ListTable>
+      <p class='annotation'>The list-table component attempts to be as flexible as possible. For this reason, <code>t.body</code> does not provide the typical <code>tr</code> element. It's sometimes desired to have multiple elements per record.</p>
+      `,
+    context: {
+      injectRoutedController: injectRoutedController(
+        Controller.extend({
+          queryParams: ['sortProperty', 'sortDescending'],
+          sortProperty: 'name',
+          sortDescending: false,
+        })
+      ),
+
+      sortedShortList: computed('controller.sortProperty', 'controller.sortDescending', function() {
+        const sorted = productMetadata.sortBy(this.get('controller.sortProperty') || 'name');
+        return this.get('controller.sortDescending') ? sorted.reverse() : sorted;
+      }),
+    },
+  };
+};
+
+export const Pagination = () => {
+  return {
+    template: hbs`
+      <h5 class="title is-5">Table pagination</h5>
+      <ListPagination @source={{longList}} @size={{5}} @page={{controller.currentPage}} as |p|>
+        <ListTable @source={{p.list}} @class="with-foot" as |t|>
+          <t.head>
+            <th class="is-1">Rank</th>
+            <th>City</th>
+            <th>State</th>
+            <th>Population</th>
+            <th>Growth</th>
+          </t.head>
+          <t.body @key="model.rank" as |row|>
+            <tr>
+              <td>{{row.model.rank}}</td>
+              <td>{{row.model.city}}</td>
+              <td>{{row.model.state}}</td>
+              <td>{{row.model.population}}</td>
+              <td>{{format-percentage row.model.growth total=1}}</td>
+            </tr>
+          </t.body>
+        </ListTable>
+        <div class="table-foot">
+          <nav class="pagination">
+            <span class="bumper-left">U.S. City population and growth from 2000-2013</span>
+            <div class="pagination-numbers">
+              {{p.startsAt}}&ndash;{{p.endsAt}} of {{longList.length}}
+            </div>
+              <p.prev @class="pagination-previous"> &lt; </p.prev>
+              <p.next @class="pagination-next"> &gt; </p.next>
```

Comment:
These are using “angle bracket invocation”, the new way to instantiate components. The `@` is how you differentiate between true HTML attributes vs properties being passed to the component. `class` is a confusing one because it’s sort of both, but this follows the pattern [in the documentation](https://guides.emberjs.com/release/reference/). (I generated it using [this codemod](https://github.com/ember-codemods/ember-angle-brackets-codemod).)

Reply:
Hmm yeah, just seems strange that as a component user you have to set these here, it feels even weirder as I have no way of knowing that a `@class` is needed as opposed to just a `class`. I'm guessing you trying to change as little as possible app code here and just getting storybook working so this is just chatter more than anything.

Out of interest, I just looked at the `list-pagination > list-pager` components, looks like they are just `<a href="">`s? Also, looks like these classes are important for the component to display properly:

https://github.com/hashicorp/nomad/blob/5884f6250c58e961a965b1bb90d4f79a8b91c239/ui/app/styles/core/pagination.scss#L10

I know all this is historical so just a note for future maybe, but I think overall it feels like these classes could go inside the component maybe? I dunno, just thinking out loud. Again fully understand you are just moving things to storybook here so little change is good, so don't feel the need to change anything!

Reply:
It's been over two years since I wrote that code (wat, how???) but I think I wanted to make `list-pagination` as generic as possible which is why the class names aren't embedded. As for `@class` vs. `class` this instance is particularly interesting since `list-pager` is tagless. It only renders an anchor with a query param href if it is "visible" (e.g., first page has no prev link). 

So to add to the semantics, using plain `class` seems incorrect because it implies that `list-pager` is an element and not fragment. Pretty small thing to yak shave though. And as you point out, a yak to shave at another time and place!